### ISSUE TITLE
big, large distributed GLM - Performance

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#3091]
http://support.sas.com/resources/papers/HPGENSELECTscalability.pdf
refers to http://blog.revolutionanalytics.com/2012/10/allstate-big-data-glm.html

150 million observations and 70 degrees of freedom

based on SAS technical paper, exog has 70 column, i.e. df_model = 70 or 69

**Comment by josef-pkt** [comment]
(just a reference that I found by chance in the SAS documentation)

**Comment by thequackdaddy** [comment]
@josef-pkt That's a funny paper. Hadn't seen it before. SAS seems very annoyed at them, which is good.

If you were ever to massively scale statsmodels GLM, how would you go about it? The limitations (in my view) are...

* statsmodels doesn't have very good support for handing large on-disk tables
* patsy, while nifty, doesn't seem to work well on my machine when you have millions of records. It can't convert categoricals to a design matrix in parallel (as far as I can tell)
* Internal operations, such as pinv, could be done on-disk and distributed. I have no clue what that will look like.

I noticed that Leland Bybee has a rather impressive looking PR dealing with distributed fits, but it appears that it is splitting the data and running sepaeate models on each split, not distributing some of the internal operations.

My big workaround at the moment has been learning bcolz, limiting my models to 5 million records, and getting more RAM. It's still a little rough, but works suitably. I've found that I can convert the bcolz ctable to a design matrix in 10-20 seconds including the time to do some munging. The actual fit takes about 90 seconds on my laptop.

Besides me, who else would use statsmodels with millions of records?

**Comment by josef-pkt** [comment]
I found it fun to see how SAS is trying to keep up, given that it's traditionally designed to work for large datasets. It looks like an interesting usecase for when we want to joing in the fun/fray.
But, I didn't try to figure out what algorithms they are using for distributed computation.

> Besides me, who else would use statsmodels with millions of records?

I think the users will show up once we provide the necessary tools given the popularity of python in datascience and given the need of traditional science to handle larger datasets.

However, one problem right now is that we don't have any larger/biggish data contributors or experts especially for the data handling part, and how to implement the communication. (I know very little, and I'm not really interested in figuring out the details myself.) Leland's GSOC will provide some of the initial infrastructure or patterns.

> it appears that it is splitting the data and running sepaeate models on each split, not distributing some of the internal operations.

The point of this approach is to require very little communication, only one or two rounds. There are similar approaches that work with aggregating separate estimation results that improve on simple aggregation of distributed results. From the design this is very flexible from parallel or out of core computing on a single machine to fully distributed computing.

Out-of-core computing with on-disk data can avoid the aggregation loss at the cost of requiring more rounds of communication, i.e. running through the data several or many times.
(If irls for GLM only takes 10 or so iterations, then using an out-of-core WLS would still be a good choice.)

**Comment by moody-marlin** [comment]
I just came across this issue and wanted to point you to a new effort for large-scale GLM using [dask](https://github.com/dask/dask), which we've dubbed [dask-glm](https://github.com/dask/dask-glm).  Currently, it's structured as a collection of optimization routines; the API is [in development](https://github.com/dask/dask-glm/issues/11) and should have a working PR soon.

Because it's built on dask, most of the scheduler / distributed aspects are hidden from the algorithm developer (so you don't have to dig deeply into the communication aspects @josef-pkt).  If you look at the actual [algorithm code](https://github.com/dask/dask-glm/blob/master/dask_glm/algorithms.py) it reads *a lot* like standard numpy code, but it operates multi-threaded or on a distributed cluster, without any change.

For some rough comparisons to the article you posted, initial profiling results on a cluster (https://github.com/dask/dask-glm/issues/26) suggest that on well-conditioned problems of size 100 million with 4 columns, run times were around 1 minute.  The obvious caveats apply: well-conditioned simulated data, no data-preprocessing involved, etc.  

Those results were run in a distributed environment, but the package works the exact same on your local machine; I ran Newton's method on my Mac (16 BG Memory / 2.5 GHz processor) on a (simulated) Logistic problem of size 10 million x 70 columns, and got convergence in ~1 minute.

If you're interested in looking into it, there are a number of open issues that could benefit from some broader community input:
- API structure (https://github.com/dask/dask-glm/issues/11)
- more profiling / benchmarking (https://github.com/dask/dask-glm/issues/26)
- Robustness / Accuracy tests (https://github.com/dask/dask-glm/issues/9, https://github.com/dask/dask-glm/issues/7)
- there are probably many more things we'll learn when people start using it more

cc: @thequackdaddy @mrocklin

**Comment by mrocklin** [comment]
Just adding a +1 to @moody-marlin 's comment.  

I'm quite happy to think about communication and parallelism issues.  I'm obviously biased here, but I think that dask is now at a pretty good place to solve this problem well.  It would be great to have more input on the mathematical and algorithmic side from people with more hands-on experience in this domain.

**Comment by josef-pkt** [comment]
@moody-marlin @mrocklin  I'm looking forward to the results. I looked at dask-glm a few times in the last month (since I saw it mentioned in one of the blog posts). I didn't try yet to figure out how it works, because it looked like a moving target.

In general I'm a bit doubtful whether rewriting the optimization routines (beyond maybe WLS, and maybe a descend method if the number of variables is huge) is the right approach. optimizers like lbfgsb only work with aggregated data which shouldn't require memory that is too large.

related question for this kind of approaches: Is there a measure of how much communication goes on until a solution is found?
for example, if dask abstracts away the distributional aspect, how can we evaluate algorithms that might differ in the amount of communication that they require, and how much performance loss might we get if the algorithm is inefficient in terms of communication?

**Comment by mrocklin** [comment]
@josef-pkt the profile results in [this issue](https://github.com/dask/dask-glm/issues/26) may interest you.  (click on one of the links corresponding to an algorithm, then select the zoom tool in the upper right of the page and zoom in)

Communication costs are highlighted in red, and don't seem to be a major performance issue for many of these algorithms.  Far more pressing at the moment is reducing scheduler overhead (which we've improved since those measurements were taken).

**Comment by mrocklin** [comment]
That being said, it could be that other algorithms are more appropriate to run at scale.  We've certainly had more success with algorithms like ADMM, which seem to be designed to not require much communication.

Are there other approaches that you would suggest?

**Comment by josef-pkt** [comment]
About the optimizers: my thought was mainly that if you use bfgs or newton, then reimplementing might not bring much (without having looked at the details).
One question for us is whether we can retrofit the existing classes (i.e. make a copy of a class and convert to dask/out-of-memory computation for everything that's possible, and strip some parts that might not be easily available.  My guess is that is easier if all operations on full data are in dask (or something like it) and the reduced part is in-memory, including the overall optimization. (I'm thinking more for problems that require out of core but in one computer where communication costs are still relatively low.)

I haven't looked at your ADMM, but IIUC, then it is related to what I was reading by Boyd last year. It looked to me like the most promising candidates for full optimization with reduced communication.

Another approach when communication is expensive, is to split up the overall estimation parts into subsample and subproblems, optimize them individually and then combine the results. We had a GSOC project last year with a first (experimental) version merged into master #3396
This doesn't do full optimization, because we don't solve the full combined optimization problem. 
Other versions of this that I read about can use more, e.g. a second round of communication to improve the approximation.

I haven't look much at this problem in the last year, my own focus has shifted to large sparse problems where we have a large number of categorical variables for example with survey data, that we are able to process in memory or on disk but where we try to keep the memory low by exploiting the sparse structure of the categorical variables.
e.g. #2710

I don't have a clear overview over the different distributed methods. dask-glm looks the most agnostic about the distributed backend because it can reuse whatever dask provides. The sparse on disk solution is for an important special case that we currently don't cover. The distributed optimization of subsamples should be fast, but could also be combined with dask-glm by providing approximate starting values for it.

**Comment by thequackdaddy** [comment]
I had stumbled across that project a few weeks ago and it looked exciting. 

What I feel is missing (not criticism, just an idea I played with a while back but real-life timelines/demands killed the meager progress I made) is building the design matrix is a real time suck for me. I spend a lot of processing time waiting to convert data from categorical to dummy variables and using splines/transforms to manipulate numeric data. For a dataset with 5 million records, I can turn about 30 columns into a design matrix with 50 columns and fit a tweedie/log glm in about 4.5 minutes. about 90 seconds of that is actually loading/data running patsy/transforms. (On a vm with 10 1.9 ghz processors... lots of lols with that). 

Patsy is just too robust/feature-complete for me to switch to something else. Of course, it lacks parallelism in and of itself--and I'm not smart enough to figure that out... so I'll leave that challenge to someone else.

**Comment by mrocklin** [comment]
> Patsy is just too robust/feature-complete for me to switch to something else. Of course, it lacks parallelism in and of itself--and I'm not smart enough to figure that out... so I'll leave that challenge to someone else.

@thequackdaddy Dask rarely reimplements an entire other project (there aren't enough developer hours to accomplish this).  Getting people to "switch" is never a goal.  Typically instead we use that project in parallel.  How are you parallellizing Patsy now across your 10 processors?  You might want to take a look at [dask.delayed](http://dask.pydata.org/en/latest/delayed.html) which was designed for this sort of purpose.

**Comment by mrocklin** [comment]
> One question for us is whether we can retrofit the existing classes (i.e. make a copy of a class and convert to dask/out-of-memory computation for everything that's possible, and strip some parts that might not be easily available. My guess is that is easier if all operations on full data are in dask (or something like it) and the reduced part is in-memory, including the overall optimization. (I'm thinking more for problems that require out of core but in one computer where communication costs are still relatively low.)

@josef-pkt it's certainly possible to make code that is simultaneously numpy/dask.array compatible (dask-glm does this, xarray does this) but it does require some effort.  There is no magic button here.  My guess is that the first step would be to identify a few algorithms where out-of-core or distributed computing would be useful and convert them to handle dask.arrays alongside numpy arrays.  That experience would probably help inform if this is an effective use of time.

The dask-glm code could be exactly that.  (And it's external, so it's easier to experiment and be reckless with.)  There are a few algorithms in there that demonstrate proof of concept.  It would be interesting to know what would need to happen to those algorithms for them to be useful to the statsmodels community.

> @moody-marlin @mrocklin I'm looking forward to the results. I looked at dask-glm a few times in the last month (since I saw it mentioned in one of the blog posts). I didn't try yet to figure out how it works, because it looked like a moving target.

Arguably now is the right time for statsmodels devs to get involved, so that they can help to direct the target to a useful place.

**Comment by thequackdaddy** [comment]
> How are you parallellizing Patsy now across your 10 processors?  

I'm not!

**Comment by josef-pkt** [comment]
@lbybee @kshedden  You might also be interested in this.

@mrocklin @moody-marlin  What's the best way to handle and store results arrays, e.g one-dimensional arrays with shape nobs like fittedvalues or resid?   maybe for efficiency we could combine those into something like a data frame.
One of the main feature of statsmodels compared to scikit-learn is that we have a large amount of inference and post-estimation results. Those usually depend on the availability of residuals and similar arrays, and for inference the full hessian.
If the number of explanatory variables is not too large (i.e. large nobs but not too large k_vars), then we can keep hessian and compute inference in memory. But with very large nobs, we might have to keep nobs arrays out of memory with several separate passes through it.
My guess is that we will be able to retrofit the results classes to work directly with dask or similar, but not the model classes. At least at the beginning I would work with stripped down copies of the model classes to handle out of memory data and estimation.

**Comment by mrocklin** [comment]
> What's the best way to handle and store results arrays, e.g one-dimensional arrays with shape nobs like fittedvalues or resid? maybe for efficiency we could combine those into something like a data frame.

nobs -> number of observations?  

It's trivial to store dask.arrays into anything that supports numpy getitem/setitem slicing.  This includes HDF5 files, PyTables, Blosc, Zarr, etc., as well as just numpy arrays if these results fit nicely in memory.  In a large distributed case we could also persist things in distributed memory.  Network bandwidth and compression can both be decently high these days.

See http://dask.pydata.org/en/latest/array-api.html#create-and-store-arrays

**Comment by mrocklin** [comment]
The other option here would be to use the dask.distributed scheduler (which works fine on a single machine) which dumps things down to disk when it starts to run low on memory.  This would automatically push least-recently used chunks of data down to disk using Dask's wire protocol, which in this case would compress with blosc and then store chunks to different files on disk.  This would be dask-specific but decently speedy and would use the memory you have available.

**Comment by lbybee** [comment]
I had a chance to look over the code earlier today and this looks very interesting.  It is certainly different from the current approach with statsmodels, as mentioned above, though it may be worth noting that what we have currently does support a dask backend through joblib.  I'm going to try and spend some time playing around with this later this week and will likely have some more comments then.

**Comment by mrocklin** [comment]
Because Joblib always moves intermediate results back to the main process it will will invariably have much more communication than using Dask directly.  I can't imagine it being nearly as efficient.

**Comment by TomAugspurger** [comment]
FYI, unless someone beats me to it, I plan to start building an integration between statsmodels and the optimizers dask-glm over the next few weeks.

**Comment by mortonjt** [comment]
@TomAugspurger do you think it would be worthwhile to enable parallelism for bulk operations (i.e. running 1000 GLMs simultaneously).  

For example, in some of our analyses, we deal with high dimensional datasets with 1000+ dimensions and would like to build GLMs for each dimension.  If that is of interest, that would be really cool!

**Comment by mrocklin** [comment]
If you're looking for embarrassingly parallel workloads it could be that the current dask interfaces would suffice.  If that is the case then it's possible that a 10 minute read through [the dask.delayed docs](http://dask.pydata.org/en/latest/delayed.html) would be enough to get you up and running.
